src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java

```diff
@@ -82,16 +103,167 @@ private static FileSystem createFileSystem(URI uri, Configuration conf)
         return fs;
     }
 
-    @SuppressWarnings("unchecked")
-    private static Map<Key, FileSystem> getPrivateMap(FileSystem.Cache cache)
+    @Override
+    synchronized void remove(Key ignored, FileSystem fileSystem)
     {
-        try {
-            Field field = FileSystem.Cache.class.getDeclaredField("map");
-            field.setAccessible(true);
-            return (Map<Key, FileSystem>) field.get(cache);
+        if (fileSystem == null) {
+            return;
         }
-        catch (ReflectiveOperationException e) {
-            throw new AssertionError(e);
+        FileSystemKey key = null;
+        for (Entry<FileSystemKey, FileSystemHolder> entry : map.entrySet()) {
+            if (fileSystem.equals(entry.getValue().getFileSystem())) {
+                key = entry.getKey();
+            }
+        }
+        if (key != null) {
+            map.remove(key);
+        }
+    }
+
+    @Override
+    synchronized void closeAll()
+            throws IOException
+    {
+        for (FileSystemHolder fileSystemHolder : map.values()) {
+            fileSystemHolder.getFileSystem().close();
+        }
+        map.clear();
+    }
+
+    @Override
+    synchronized void closeAll(boolean onlyAutomatic)
+            throws IOException
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    synchronized void closeAll(UserGroupInformation ugi)
+            throws IOException
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    private static FileSystemKey createFileSystemKey(URI uri, UserGroupInformation userGroupInformation, long unique)
+    {
+        String scheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase();
+        String authority = uri.getAuthority() == null ? "" : uri.getAuthority().toLowerCase();
+        String realUser;
+        String proxyUser = null;
```

Comment:
Remove the initialization here and do it below, so it's clear when we set it to null.